### PR TITLE
fs cache: continue without exception when getting no disk space left error

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Continue without exception when running out of disk space when using filesystem cache.

### CODE REVIEW

File: src/Disks/IO/CachedReadBufferFromRemoteFS.cpp [hunk 1]
```diff
@@ -691,6 +694,34 @@ bool CachedReadBufferFromRemoteFS::updateImplementationBufferIfNeeded()
...
+    Stopwatch watch(CLOCK_MONOTONIC);
+
+    try
+    {
+        file_segment.write(data, size, offset);
+    }
+    catch (ErrnoException & e)
+    {
+        int code = e.getErrno();
+        if (code == /* No space left on device */28 || code == /* Quota exceeded */122)
```
Comment (rschu1ze) [inline src/Disks/IO/CachedReadBufferFromRemoteFS.cpp:708]: Perhaps a little bit more idiomatic: `if (code == /* No space left on device */ ENOSPC || code == /* Quota exceeded */ EDQUOT)`

EDIT: On Darwin, macro `EDQUOT` isn't available ... using numerical codes is okay then.

File: src/Disks/IO/CachedReadBufferFromRemoteFS.cpp [hunk 2]
```diff
@@ -691,6 +694,34 @@ bool CachedReadBufferFromRemoteFS::updateImplementationBufferIfNeeded()
...
+        {
+            LOG_INFO(log, "Not enough disk space to write cache, will skip cache download. ({})", e.displayText());
```
Comment (rschu1ze) [inline src/Disks/IO/CachedReadBufferFromRemoteFS.cpp:710]: Maybe a bit shorter: `Insert into cache skipped due to insufficient disk space. ({})` ?

File: src/Disks/IO/CachedReadBufferFromRemoteFS.cpp [hunk 3]
```diff
@@ -840,33 +871,34 @@ bool CachedReadBufferFromRemoteFS::nextImplStep()
...
+                if (success)
+                {
+                    assert(file_segment->getDownloadOffset() <= file_segment->range().right + 1);
+                    assert(
+                        std::next(current_file_segment_it) == file_segments_holder->file_segments.end()
+                        || file_segment->getDownloadOffset() == implementation_buffer->getFileOffsetOfBufferEnd());
+                }
+                else
+                {
+                    assert(file_segment->state() == FileSegment::State::PARTIALLY_DOWNLOADED_NO_CONTINUATION);
```
Comment (rschu1ze) [inline src/Disks/IO/CachedReadBufferFromRemoteFS.cpp:889]: I wonder where `file_segment->state()` is set to `FileSegment::State::PARTIALLY_DOWNLOADED_NO_CONTINUATION`? At least `writeCache()` (l. 879) doesn't do that.

**Review by rschu1ze — APPROVED** [review]

File: src/Disks/IO/CachedReadBufferFromRemoteFS.cpp (hunk 3 above)
Comment (kssenii) [reply src/Disks/IO/CachedReadBufferFromRemoteFS.cpp:889]: It does. In `file_segment->write()` there is `setDownloadFailed()` which updates this state in try catch.

**Review by rschu1ze — APPROVED** [review]

**Review by rschu1ze — APPROVED** [review]

**Pull request merged**